### PR TITLE
[WIP] Improve 32 bit build.

### PR DESCRIPTION
**Pull request by bernhardu**
Sorry for the delay. This PR is related to #2597.

The given 6 commits help to build current git in a debian stable VM.
It passes all tests except these:
```
The following tests FAILED:
        898 - nested_detach_wait (Failed)
        899 - nested_detach_wait-no-syscallbuf (Failed)
        1118 - nested_detach (Failed)
        1119 - nested_detach-no-syscallbuf (Failed)
```
From my point of view it looks like the child process started by the nested rr has no valid instruction pointer.

I would be glad to get some feedback to change some of the hacks to proper patches.

### CODE REVIEW

File: src/ExtraRegisters.cc [hunk 1]
```diff
@@ -144,6 +144,9 @@ static uint64_t xsave_features(const vector<uint8_t>& data) {
                                                   xsave_header_offset);
 }
 
+#if defined (__i386__)
+  typedef long double __uint128_t;
```
Comment (rocallahan) [inline src/ExtraRegisters.cc:148]: This doesn't look right.

I think we can just replace `sizeof(__uint128_t)` with `16`.

File: src/kernel_abi.h [hunk 2]
```diff
@@ -2103,6 +2103,9 @@ struct ARM64Arch : public GenericArch<SupportedArch::aarch64, WordSize64Defs> {
   };
   typedef struct user_pt_regs user_regs_struct;
 
+#if defined (__i386__)
+  typedef long double __uint128_t;
+#endif
```
Comment (rocallahan) [inline src/kernel_abi.h:2108]: This case is a bit more painful.

`long double` isn't right. I think instead something like
```
typedef struct __uint128_t {
  uint64_t parts[2];
};
```

File: src/record_signal.cc [hunk 3]
```diff
@@ -166,9 +166,14 @@ static bool try_grow_map(RecordTask* t, siginfo_t* si) {
     LOG(debug) << "try_grow_map " << addr << ": address would be in guard page";
     return false;
   }
-  struct rlimit64 stack_limit;
   remote_ptr<void> limit_bottom;
+#if not defined (__i386__)
```
Comment (rocallahan) [inline src/record_signal.cc:170]: Use the `!` operator please ... I didn't even know you can write `not` here

File: src/Task.cc [hunk 4]
```diff
@@ -3435,7 +3435,11 @@ void Task::dup_from(Task *other) {
       remote_this.syscall(syscall_number_for_fchdir(this->arch()), child_fd);
       remote_this.syscall(syscall_number_for_close(this->arch()), child_fd);
     }
+#if not defined (__i386__)
     struct prctl_mm_map map;
+#else
+    struct prctl_mm_map map = {};
```
Comment (rocallahan) [inline src/Task.cc:3441]: Just do this unconditionally.

File: src/record_signal.cc (hunk 3 above)
Comment (rocallahan) [reply src/record_signal.cc:170]: Actually, flip the order of the if/else blocks so you don't need to use "not" at all.

**Comment by bernhardu** [comment]
Thank you for the review.
If I remember correctly I just have to rewrite the commits in my local tree in the same branch and `git push --force`?

File: src/ExtraRegisters.cc (hunk 1 above)
Comment (bernhardu) [reply src/ExtraRegisters.cc:148]: So it should read this:
```
        ((regno - DREG_V0) * 16/*sizeof(__uint128_t)*/), 16/*sizeof(__uint128_t)*/);
```
Or would be this better:
```
        ((regno - DREG_V0) * 16), 16); // type __uint128_t is not defined at i386.
```

**Comment by khuey** [comment]
Yes force pushing your branch is fine.

File: src/ExtraRegisters.cc (hunk 1 above)
Comment (rocallahan) [reply src/ExtraRegisters.cc]: This:
```
((regno - DREG_V0) * 16), 16);
```

File: src/ExtraRegisters.cc [hunk 5]
```diff
@@ -155,7 +155,7 @@ size_t ExtraRegisters::read_register(uint8_t* buf, GdbRegister regno,
     RegData reg_data;
     if (DREG_V0 <= regno && regno <= DREG_V31) {
       reg_data = RegData(offsetof(ARM64Arch::user_fpsimd_state, vregs[0]) +
-        ((regno - DREG_V0) * sizeof(__uint128_t)), sizeof(__uint128_t));
+        ((regno - DREG_V0) * 16/*sizeof(__uint128_t)*/), 16/*sizeof(__uint128_t)*/);
```
Comment (rocallahan) [inline src/ExtraRegisters.cc:158]: Not this :-)

**Comment by bernhardu** [comment]
Hello, I tracked down another issue with copy_file_range_all similar to the first off_t/off64_t patch.
Can I still add it to this PR?

**Comment by khuey** [comment]
Yes.

**Comment by bernhardu** [comment]
The last patch makes the segfault of the child-child disappear, but that's still not enough to make the nested_detach test succeed.
The child-child now silently vanishes ... Are there in that early state some other debugging options?

**Comment by rocallahan** [comment]
Please file a separate issue for the `nested_detach` test failure.

**Pull request merged**